mime.go

```diff
@@ -107,8 +120,96 @@ func (mh *mimeHandler) initJPEGEncoder(frame *C.AVFrame) error {
 	return nil
 }
 
+func (mh *mimeHandler) convertYUYV(frame *C.AVFrame) ([]byte, camera.ImageMetadata, error) {
+	if mh.yuyvSwsCtx == nil || frame.width != mh.yuyvFrame.width || frame.height != mh.yuyvFrame.height {
+		if err := mh.initYUYVContext(frame); err != nil {
+			return nil, camera.ImageMetadata{}, err
+		}
+	}
+	if mh.yuyvSwsCtx == nil {
+		return nil, camera.ImageMetadata{}, errors.New("failed to create YUYV converter")
+	}
+	if mh.yuyvFrame == nil {
+		return nil, camera.ImageMetadata{}, errors.New("failed to create yuyv destination frame")
+	}
+	res := C.sws_scale(
+		mh.yuyvSwsCtx,
+		(**C.uint8_t)(unsafe.Pointer(&frame.data[0])),
+		(*C.int)(unsafe.Pointer(&frame.linesize[0])),
+		0,
+		frame.height,
+		(**C.uint8_t)(unsafe.Pointer(&mh.yuyvFrame.data[0])),
+		(*C.int)(unsafe.Pointer(&mh.yuyvFrame.linesize[0])),
+	)
+	if res < 0 {
+		return nil, camera.ImageMetadata{}, newAvError(res, "failed to convert frame to YUYV")
+	}
+	yuyvDataSize := int(mh.yuyvFrame.width) * int(mh.yuyvFrame.height) * yuyvBytesPerPixel
+	header := packYUYVHeader(int(mh.yuyvFrame.width), int(mh.yuyvFrame.height))
+	// Preallocate the final slice with the combined size of the header and the image data
+	yuyvPacket := make([]byte, len(header)+yuyvDataSize)
+	copy(yuyvPacket[0:], header)
+	// Copy the YUYV data directly into the preallocated slice
+	C.memcpy(unsafe.Pointer(&yuyvPacket[len(header)]), unsafe.Pointer(mh.yuyvFrame.data[0]), C.size_t(yuyvDataSize))
+
+	return yuyvPacket, camera.ImageMetadata{
+		MimeType: mimeTypeYUYV,
+	}, nil
+}
+
+func (mh *mimeHandler) initYUYVContext(frame *C.AVFrame) error {
+	mh.mu.Lock()
+	defer mh.mu.Unlock()
+	mh.logger.Info("creating YUYV sws cosntext with frame size: ", frame.width, "x", frame.height)
+	if mh.yuyvSwsCtx != nil {
+		C.sws_freeContext(mh.yuyvSwsCtx)
+	}
+	if mh.yuyvFrame != nil {
+		C.av_frame_free(&mh.yuyvFrame)
+	}
+	mh.yuyvFrame = C.av_frame_alloc()
+	if mh.yuyvFrame == nil {
+		return errors.New("failed to allocate YUYV frame")
+	}
+	mh.yuyvFrame.width = frame.width
+	mh.yuyvFrame.height = frame.height
+	mh.yuyvFrame.format = C.AV_PIX_FMT_YUYV422
```

Comment:
do we ever need to care about yuyv420 differences here?

Reply:
The output of the sws_scale operation here is always `AV_PIX_FMT_YUYV422`. We can handle multiple input formats (`YUV420P` or `YUVJ420P`) which is automatically handled in the decoder to avframe pool pipeline.